### DISCUSSION TITLE
Non-deterministic behavior / race condition in hooks

### DISCUSSION THREAD

**Discussion by ssyberg** [payloadcms/payload#2785] [Q&A]
### Link to reproduction

N/A

### To Reproduce

This is a fairly complex set up and I haven't been able to reproduce it in a trivial example so I'm hoping something jumps out to whoever reviews this. I've encountered this bug a number of times over the last few months and because it's non-deterministic it's very hard to reliably reproduce. 

### Describe the Bug

I have two examples of this, one was where I setting a `console.log` within a beforeValidate hook would change the outcome of the validation (see this discord thread: https://discord.com/channels/967097582721572934/1091785894136139808/1091785894136139808).

More recently I've lost a few hours trying to debug this beforeChange hook and similarly hooks are firing and data is available non-deterministically:

```
name: 'path',
unique: true,
admin: {
    hidden: true,
},
hooks: {
    beforeChange: [
        ({ data }) => {
            if (data.slug && data.parent && data.parent.value) {
                return `${data.parent.value}/${data.slug}`;
            }
        },
    ],
},
```

Some important context:
- `slug` is being set in a `beforeValidate` hook
- `parent` is being set by a async method on `defaultValue`

This seems to reliably work in the admin UI, but never works on a local API `create`. When setting breakpoints in the above beforeChange hook I see data.slug is set correctly by it's beforeValidate hook, but data.parent is undefined.

FWIW I've tried different combos of beforeChange, beforeValidate, afterChange to address this and nothing seems to reliably solve it.

One important data point, adding a sleep completely solves this which makes me pretty confident it's a race condition:
```
hooks: {
    beforeChange: [
        async ({ data }) => {
            await new Promise((resolve) => setTimeout(resolve, 100));
        },
        ({ data }) => {
            if (data.slug && data.parent && data.parent.value) {
                return `${data.parent.value}/${data.slug}`;
            }
        },
    ],
},
```

### Payload Version

1.8.3 (but have noticed this since 1.6.x)

**Comment by swenzel-arc** [comment]
If your _field_ hooks depend on values of other fields that are themselves modified by their own field hooks, it might be wiser to merge all that logic into a collection level hook so you have full control of the order of execution and can edit all fields at the same place.

**Comment by ssyberg** [comment]
> If your _field_ hooks depend on values of other fields that are themselves modified by their own field hooks, it might be wiser to merge all that logic into a collection level hook so you have full control of the order of execution and can edit all fields at the same place.

That is a solution, though for my specific implementation this is occurring on an importable field group, what you're suggesting would mean importing the field group and setting up a collection hook on every collection that uses this which feels a bit brittle and not properly encapsulated.

That said I'd still expect inter-hook behavior to be deterministic, this still feels like a bug to me.

**Comment by jmikrut** [comment]
Let me chime in here quick with some additional info.

Field hooks are intentionally all run _**in parallel**_ to one another for each step. So your `parent` field is calculating its `defaultValue` and your `path` field is attempting to read that value at the same time, in parallel (within the `beforeChange` lifecycle event / hook). 

This is a performance optimization as well as a side-effect of the way that JavaScript works. The properties of field data are just keys of a plain old JavaScript object, which do not have intrinsic / reliable order. So even though we do map through the properties of your field data, the order is not predictable.

If you do need predictability in certain fields' calculations, then I would view the different hooks as the only reliable way to guarantee hook order. Or, as @swenzel-arc mentioned, just use a collection hook instead. 

What would the alternative be here? I am certainly not opposed to building some type of "field hook ordering" system, but that would be more of a new feature and less of an issue with Payload as it stands now.

What do you think? I'll convert this to a discussion but if I have misunderstood something, and there is indeed a bug with Payload here, I'd be happy to re-open as an issue.

**Reply by ssyberg** [reply]
Part of the confusion might just be a lack of documentation on the hooks / life cycle. It makes sense that all `beforeValidate` hooks are called in parallel, but it's not clear when the `defaultValue` callbacks are run. I would still expect `beforeValidate` hooks to all complete before the `beforeChange` hooks are called, but maybe that's what you're already saying and that it's just a cascading effect between the `beforeValidate` and the `defaultValue` callback?

**Reply by ssyberg** [reply]
@jmikrut just bumping this, can you clarify two things:
(1) when are defaultValue callbacks run? I would have expected this to occur **before** beforeValidate hooks are run
(2) can you confirm that each level of field hook is run in parallel on all fields, but that the beforeValidate hook of field A is guaranteed to run before the beforeChange hook of field B?

**Reply by ssyberg** [reply]
I think this is still a bug if for no other reason than we have divergent behavior between what this sentence means depending on a local API `create` and a admin dashboard `create` and that smells funny to me:

> The data passed to update the document within create and update operations, and the full document itself in the afterRead hook.

It's my understanding that in the admin dashboard you're guaranteed that defaultValue is going to have executed and is set on `data` in all hooks if using the dashboard, but it's non-deterministic if it's using the local API.

**Reply by jmikrut** [reply]
Here are some answers:

The `beforeValidate` hook runs BEFORE the `defaultValue` is calculated. Default values are calculated explicitly in the `beforeChange` hook, before the validation functions are run. So the order goes like this: 

**The `beforeValidate` event runs** - and within, the following happens:

1. Field data "sanitization" - AKA do things like convert `'true'` to `true` if the field is a `checkbox`, sanitize incoming values as `number` if the field type is `number`, etc.
2. Field `beforeValidate` hooks run
3. Field-level access control is run

**Next, the `beforeChange` event runs, with the following:**

1. Run condition (to determine if validation should or should not take place)
2. Merge original document data into incoming data
3. Compute default values for undefined fields (NOTE that this happens BEFORE validation, but within this lifecycle event
4. Execute `beforeChange` field hooks
5. Validate data
6. Transform data for storage if necessary (point field, etc)
7. "Unflatten" locales -> merge in all other locales on a doc, if applicable

> can you confirm that each level of field hook is run in parallel on all fields, but that the beforeValidate hook of field A is guaranteed to run before the beforeChange hook of field B?

Yes, ALL events within `beforeValidate` will finish before _anything_ begins in `beforeChange`.

However, a quick nitty-gritty clarification-it's not necessarily ALL fields at once that run in parallel. It's hierarchical for each nested level of a document's fields. For example, all top-level fields do indeed run at the same time. But after those are finished, each structure of nested fields (groups, array / block rows, named tab) run at the same time. Repeat until the lowest-point of the document.

> when are defaultValue callbacks run? I would have expected this to occur before beforeValidate hooks are run

As you see above, default values are calculated within the `beforeChange` event itself, before the field hooks themselves fire. 

I think much of the confusion here is because as you see, within the `beforeChange` event, indeed there are a few things happening potentially in parallel - including default values, the `beforeChange` hooks themselves, and validation.

This is all built very deliberately to optimize performance and reduce the amount of looping through data structures that needs to happen on the server. We could have built out separate events for each of the above actions one after another, for example, first run all conditions, then merge data, then compute default values,  then execute `beforeChange`, then execute validation, etc. - - but that is a lot of looping and a lot of blocking. It worked out best by quite a bit to run all of this in a single loop per field.

Last note - some of these steps get executed in a different context / scenario in the admin UI, because there, the requirements are dramatically different.

Examples of admin-side differences:

1. Validation runs as you change the document, constantly
2. Conditions run constantly as well
3. Default values are calculated only on first load of the document, _**not**_ before saving
4. Locales are irrelevant admin-side

Does this answer your question?

**Reply by ssyberg** [reply]
I think this still speaks to a odd divergence between admin and API behavior:

> The beforeValidate hook runs BEFORE the defaultValue is calculated

For example in the dashboard the defaultValue of a select or radio will be preselected on the form on load and as such it's executed before the beforeValidate hook.

**Reply by ssyberg** [reply]
☝🏼 and I might add it's not obvious how to write hooks/defaults that play nicely in both scenarios.

**Reply by jmikrut** [reply]
> For example in the dashboard the defaultValue of a select or radio will be preselected on the form on load and as such it's executed before the beforeValidate hook.

Well, remember that the beforeValidate hook doesn’t run at all in the browser though. So intrinsically the context is different. 

In the admin UI, however, default values are indeed calculated before running validate functions - same as the server. So there is parity there. 

> and I might add it's not obvious how to write hooks/defaults that play nicely in both scenarios.

totally agreed with this and improving the docs for the Payload “lifecycle” is work that we are about to take on. You can expect more from us there shortly and that should help clarify a lot of this!

**Reply by jmikrut** [reply]
Oh also – we are about to release a feature that allows you to build “context“ for the hooks lifecycle, which will allow you to wait on certain actions to complete before running other hook actions. This might not be the most simple application of that new feature, and you might be better off trying to simplify rather than leverage the new hooks context here specifically, but thought I’d throw that out 👍

**Reply by ssyberg** [reply]
Cool thank you for all the responses! I am still curious what the downside is to running the defaultValue callback before any other hooks since that is still effectively what happens when it's "run in the browser"? From the servers perspective, I think it would make sense for the defaultValue to be treated as if it was coming from a PATCH/POST and that would address this behavior difference...
